Re-running failing test fails with classloader errors

Re-occurrence of #46048. Originally reported (along with an unrelated issue) in https://github.com/quarkusio/quarkus/issues/50740.

```
org.junit.jupiter.engine.execution.ConditionEvaluationException: Failed to evaluate condition [io.quarkus.test.junit.QuarkusTestExtension]: Internal error: Test class was loaded with an unexpected classloader or the thread context classloader was incorrect.
  at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:197)
  at java.base/java.util.stream.ReferencePipeline$2$1.accept(ReferencePipeline.java:179)
  at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:197)
  at java.base/java.util.stream.ReferencePipeline$2$1.accept(ReferencePipeline.java:179)
  at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:197)
  at java.base/java.util.ArrayList$ArrayListSpliterator.tryAdvance(ArrayList.java:1685)
  at java.base/java.util.stream.ReferencePipeline.forEachWithCancel(ReferencePipeline.java:129)
  at java.base/java.util.stream.AbstractPipeline.copyIntoWithCancel(AbstractPipeline.java:527)
  at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:513)
  at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:499)
  at java.base/java.util.stream.FindOps$FindOp.evaluateSequential(FindOps.java:150)
  at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
  at java.base/java.util.stream.ReferencePipeline.findFirst(ReferencePipeline.java:647)
  at java.base/java.util.ArrayList.forEach(ArrayList.java:1596)
Caused by: java.lang.IllegalStateException: Internal error: Test class was loaded with an unexpected classloader or the thread context classloader was incorrect.
  at io.quarkus.test.junit.AbstractJvmQuarkusTestExtension.evaluateExecutionCondition(AbstractJvmQuarkusTestExtension.java:204)
  ... 14 more
Caused by: java.util.ServiceConfigurationError: io.smallrye.config.SmallRyeConfigFactory: io.quarkus.runtime.configuration.QuarkusConfigFactory not a subtype
  at java.base/java.util.ServiceLoader.fail(ServiceLoader.java:593)
  at java.base/java.util.ServiceLoader$LazyClassPathLookupIterator.hasNextService(ServiceLoader.java:1244)
  at java.base/java.util.ServiceLoader$LazyClassPathLookupIterator.hasNext(ServiceLoader.java:1273)
  at java.base/java.util.ServiceLoader$2.hasNext(ServiceLoader.java:1309)
  at java.base/java.util.ServiceLoader$3.hasNext(ServiceLoader.java:1393)
  at io.smallrye.config.SmallRyeConfigProviderResolver.getFactoryFor(SmallRyeConfigProviderResolver.java:102)
  at io.smallrye.config.SmallRyeConfigProviderResolver.getConfig(SmallRyeConfigProviderResolver.java:78)
  at io.smallrye.config.SmallRyeConfigProviderResolver.getConfig(SmallRyeConfigProviderResolver.java:66)
  at org.eclipse.microprofile.config.ConfigProvider.getConfig(ConfigProvider.java:85)
  at io.quarkus.test.junit.AbstractJvmQuarkusTestExtension.evaluateExecutionCondition(AbstractJvmQuarkusTestExtension.java:162)
```

I believe we're seeing this again because something in Surefire 3.5.4 exposes the problem, and we upgraded to that version in September (https://github.com/quarkusio/quarkus/pull/50092). Unfortunately, that update was needed to fix #50076, so if we downgraded to 3.5.3, it would re-expose that problem. 

I've confirmed that https://github.com/apache/maven-surefire/pull/863 is what's causing us problems, so I'll have a look to see if I can make our test launching co-exist with that change. I don't see many options for making it work, unfortunately. Alternatively, we could ask Surefire to revert the problem PR.

--- a/integration-tests/maven/src/test/resources-filtered/projects/test-flaky-test-multiple-profiles/pom.xml
+++ b/integration-tests/maven/src/test/resources-filtered/projects/test-flaky-test-multiple-profiles/pom.xml
@@ -89,6 +89,12 @@
           </execution>
         </executions>
       </plugin>
+      <!-- Always use the same version of surefire as the main Quarkus build -->
+      <plugin>
+          <groupId>org.apache.maven.plugins</groupId>
+          <artifactId>maven-surefire-plugin</artifactId>
+          <version>${version.surefire.plugin}</version>
+      </plugin>
     </plugins>
   </build>
   <profiles>

--- a/integration-tests/maven/src/test/resources-filtered/projects/test-flaky-test-single-profile/pom.xml
+++ b/integration-tests/maven/src/test/resources-filtered/projects/test-flaky-test-single-profile/pom.xml
@@ -89,6 +89,12 @@
           </execution>
         </executions>
       </plugin>
+      <!-- Always use the same version of surefire as the main Quarkus build -->
+      <plugin>
+          <groupId>org.apache.maven.plugins</groupId>
+          <artifactId>maven-surefire-plugin</artifactId>
+          <version>${version.surefire.plugin}</version>
+      </plugin>
     </plugins>
   </build>
   <profiles>

--- a/test-framework/junit5/src/main/java/io/quarkus/test/junit/classloading/FacadeClassLoader.java
+++ b/test-framework/junit5/src/main/java/io/quarkus/test/junit/classloading/FacadeClassLoader.java
@@ -267,7 +267,9 @@ public final class FacadeClassLoader extends ClassLoader implements Closeable {
 
     @Override
     public Class<?> loadClass(String name) throws ClassNotFoundException {
-        log.debugf("Facade classloader loading %s", name);
+        // Debuggers, beware: If the same class is loaded with the same FacadeClassLoader instance using Class.forName(), a cached copy will be returned,
+        // and nothing will be logged here; it will look like classloading didn't happen, but the problem is actually a stale instance
+        log.debugf("Facade classloader loading %s for the first time\n", name);
 
         if (peekingClassLoader == null) {
             throw new RuntimeException("Attempted to load classes with a closed classloader: " + this);

--- a/test-framework/junit5/src/main/java/io/quarkus/test/junit/launcher/CustomLauncherInterceptor.java
+++ b/test-framework/junit5/src/main/java/io/quarkus/test/junit/launcher/CustomLauncherInterceptor.java
@@ -4,10 +4,13 @@ import org.junit.platform.launcher.LauncherDiscoveryListener;
 import org.junit.platform.launcher.LauncherDiscoveryRequest;
 import org.junit.platform.launcher.LauncherSession;
 import org.junit.platform.launcher.LauncherSessionListener;
+import org.junit.platform.launcher.TestExecutionListener;
+import org.junit.platform.launcher.TestPlan;
 
 import io.quarkus.test.junit.classloading.FacadeClassLoader;
 
-public class CustomLauncherInterceptor implements LauncherDiscoveryListener, LauncherSessionListener {
+public class CustomLauncherInterceptor
+        implements LauncherDiscoveryListener, LauncherSessionListener, TestExecutionListener {
 
     private static FacadeClassLoader facadeLoader = null;
     // Also use a static variable to store a 'first' starting state that we can reset to
@@ -93,7 +96,6 @@ public class CustomLauncherInterceptor implements LauncherDiscoveryListener, Lau
             System.setProperty("java.util.concurrent.ForkJoinPool.common.threadFactory",
                     "io.quarkus.bootstrap.forkjoin.QuarkusForkJoinWorkerThreadFactory");
         }
-
     }
 
     private void adjustContextClassLoader() {
@@ -108,7 +110,6 @@ public class CustomLauncherInterceptor implements LauncherDiscoveryListener, Lau
 
     @Override
     public void launcherDiscoveryFinished(LauncherDiscoveryRequest request) {
-
         if (!isProductionModeTests()) {
             // We need to support two somewhat incompatible scenarios.
             // If there are user extensions present which implement `ExecutionCondition`, and they call config in `evaluateExecutionCondition`,
@@ -129,7 +130,10 @@ public class CustomLauncherInterceptor implements LauncherDiscoveryListener, Lau
 
     @Override
     public void launcherSessionClosed(LauncherSession session) {
+        clearContextClassloader();
+    }
 
+    private static void clearContextClassloader() {
         try {
             // Tidy up classloaders we created, but not ones created upstream
             // Also make sure to reset the TCCL so we don't leave a closed classloader on the thread
@@ -149,5 +153,38 @@ public class CustomLauncherInterceptor implements LauncherDiscoveryListener, Lau
         } catch (Exception e) {
             throw new RuntimeException("Failed to close custom classloader", e);
         }
+    }
+
+    /**
+     * Called when the execution of the {@link TestPlan} has started,
+     * <em>before</em> any test has been executed.
+     *
+     * <p>
+     * Called from the same thread as {@link #testPlanExecutionFinished(TestPlan)}.
+     *
+     * In continuous testing, the test plan listener seems not to be called, perhaps because of a different execution model.
+     *
+     * @param testPlan describes the tree of tests about to be executed
+     */
+    public void testPlanExecutionStarted(TestPlan testPlan) {
+        // Do nothing, but have the method here for symmetry :)
+    }
+
+    /**
+     * Called when the execution of the {@link TestPlan} has finished,
+     * <em>after</em> all tests have been executed.
+     *
+     * <p>
+     * Called from the same thread as {@link #testPlanExecutionStarted(TestPlan)}.
+     *
+     * If tests failed and are rerun by surefire, the same session will be used for all runs, so we need to get rid of the
+     * FacadeClassLoader associated with the previous run, since its app will be closed and its classloaders will all be stale.
+     *
+     * In continuous testing, the test plan listener seems not to be called, perhaps because of a different execution model.
+     *
+     * @param testPlan describes the tree of tests that have been executed
+     */
+    public void testPlanExecutionFinished(TestPlan testPlan) {
+        clearContextClassloader();
     }
 }

--- a/test-framework/junit5/src/main/resources/META-INF/services/org.junit.platform.launcher.TestExecutionListener
+++ b/test-framework/junit5/src/main/resources/META-INF/services/org.junit.platform.launcher.TestExecutionListener
@@ -1,1 +1,2 @@
 io.quarkus.test.junit.launcher.ExecutionListener
+io.quarkus.test.junit.launcher.CustomLauncherInterceptor

